### PR TITLE
Rework WarningLevel property

### LINKED ISSUE #7977: ArgumentNullException when setting same value across all configuration dimensions
## Visual Studio Version

17.2.0 Preview 3.0 [32308.589.main]

## Summary

It's possible to trigger a yellow bar while using project properties.

## Steps to Reproduce

1. Create a new .NET Console App (targeting .NET 6)
2. Open the Project Properties UI
3. Navigate to Build | Errors and warnings | Warning level
4. Click the gear to vary by configuration
5. Click the gear again to use the same value across all configurations

This produces a yellow bar for exception:

```
=====================
15/03/2022 10:50:55
Recoverable
System.AggregateException: One or more errors occurred. ---> System.ArgumentNullException: Value cannot be null.
Parameter name: unevaluatedPropertyValue
   at Microsoft.Requires.NotNull[T](T value, String parameterName)
   at Microsoft.VisualStudio.ProjectSystem.Properties.ProjectPropertiesBase.Microsoft.VisualStudio.ProjectSystem.Properties.IContextSpecificProjectProperties.SetPropertyValueInContextAsync(String propertyName, String groupLabel, String unevaluatedPropertyValue, IReadOnlyDictionary`2 dimensionalConditions)
   at Microsoft.VisualStudio.ProjectSystem.PropertyPages.PageProperty.<SetValueAtDataSourceAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.PropertyPages.PageEnumPropertyBase.<SetValueAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.VS.Query.ProjectSetUIPropertyValueActionCore.<ExecuteAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Query.ProjectSetUIPropertyValueActionBase`1.<ReceiveResultAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectUpdateBatchProvider.ProjectUpdateBatchExecutor.<>c__DisplayClass4_0.<<ExecuteActionsAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.ProjectLockService.<ExecuteWithinLockAsync>d__129.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.ProjectLockService.<ExecuteWithinLockAsync>d__129.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectUpdateBatchProvider.ProjectUpdateBatchExecutor.<ExecuteActionsAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.QueryBatchExecutorBase.<OnRequestProcessFinishedAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.QueryUpdateExecutionManager.<OnRequestProcessFinishedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ExecutableQueryDataProducer`2.InnerResultReceiver.<Microsoft-VisualStudio-ProjectSystem-Query-QueryExecution-IQueryResultReceiver<TResult>-OnRequestProcessFinishedAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectQueryDataProvider.ProjectByIdDataProducer.<SendRequestAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.MergedQueryByIdDataProducer.ProcessManager.<ProcessRequestAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ExecutableQueryDataProducer`2.<ExecuteAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ServerRuntimeExecutionEngine.<RunDataProducerAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ServerRuntimeExecutionEngine.<ExecuteExecutableAsync>d__11`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.PropertyPages.Designer.PropertyContext.<>c__DisplayClass37_0.<<SetEvaluatedValue>b__0>d.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ArgumentNullException: Value cannot be null.
Parameter name: unevaluatedPropertyValue
   at Microsoft.Requires.NotNull[T](T value, String parameterName)
   at Microsoft.VisualStudio.ProjectSystem.Properties.ProjectPropertiesBase.Microsoft.VisualStudio.ProjectSystem.Properties.IContextSpecificProjectProperties.SetPropertyValueInContextAsync(String propertyName, String groupLabel, String unevaluatedPropertyValue, IReadOnlyDictionary`2 dimensionalConditions)
   at Microsoft.VisualStudio.ProjectSystem.PropertyPages.PageProperty.<SetValueAtDataSourceAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.PropertyPages.PageEnumPropertyBase.<SetValueAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.VS.Query.ProjectSetUIPropertyValueActionCore.<ExecuteAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Query.ProjectSetUIPropertyValueActionBase`1.<ReceiveResultAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectUpdateBatchProvider.ProjectUpdateBatchExecutor.<>c__DisplayClass4_0.<<ExecuteActionsAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.ProjectLockService.<ExecuteWithinLockAsync>d__129.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.ProjectLockService.<ExecuteWithinLockAsync>d__129.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectUpdateBatchProvider.ProjectUpdateBatchExecutor.<ExecuteActionsAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.QueryBatchExecutorBase.<OnRequestProcessFinishedAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.QueryUpdateExecutionManager.<OnRequestProcessFinishedAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ExecutableQueryDataProducer`2.InnerResultReceiver.<Microsoft-VisualStudio-ProjectSystem-Query-QueryExecution-IQueryResultReceiver<TResult>-OnRequestProcessFinishedAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.QueryDataProviders.ProjectQueryDataProvider.ProjectByIdDataProducer.<SendRequestAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.MergedQueryByIdDataProducer.ProcessManager.<ProcessRequestAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ExecutableQueryDataProducer`2.<ExecuteAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ServerRuntimeExecutionEngine.<RunDataProducerAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution.ServerRuntimeExecutionEngine.<ExecuteExecutableAsync>d__11`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.PropertyPages.Designer.PropertyContext.<>c__DisplayClass37_0.<<SetEvaluatedValue>b__0>d.MoveNext()<---
```

### PR DESCRIPTION
**Pull request by tmeschter**
Fixes #7977.
Related to [AB#1459442](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1459442).

Historically, the `WarningLevel` property (corresponding to the `-warn:<n>` switch) controlled what set of warnings would be produced by the C# compiler. "0" meant produce no warnings (only errors). "1" meant produce only the most severe warnings. "2" meant produce all the warnings in 1, plus a set of less severe warnings. "3" meant all the warnings in 2 (and 1) plus additional, even less severe warnings. And so on. The default for a long time was "4" which effectively turned on all warnings the compiler could produce. The general idea was that, if need be, the user could pick a low `WarningLevel` to identify and subsequently fix the most impactful warnings in their code, and gradually ratchet the level up over time.

The scope and effective behavior of `WarningLevel` changed with .NET 5 and C# 9.0. This added a new warning level, 5, which controlled not only additional warnings added with C# 9.0 support, but _also_ certain analyzers included with the .NET 5.0 SDK. The SDK also sets the `WarningLevel` to 5 by default. Similarly, the .NET 6.0 SDK adds level 6, sets that as the default, and adds new analyzers/warnings tied to that level. The idea now is that as you adopt newer versions of the SDK you get new warnings, but have the option of turning them off by reducing the `WarningLevel`.

Along the way the logic of how `WarningLevel` is set to a default value in the SDK has become complicated. If you are targeting .NET 5.0 or later, the `WarningLevel` is computed based on the `EffectiveAnalysisLevel` property, which is itself computed from `AnalysisLevel`. If the `EffectiveAnalysisLevel` is 5.0 or greater, the SDK targets will override any `WarningLevel` set in the project in favor of specific values that correspond to the `EffectiveAnalysisLevel`. In these cases it doesn't make sense to provide controls to the user to change `WarningLevel`; the new value will be written to the project file but won't be honored, and the control will go back to the value provided by the SDK.

Here we address this by adding a hidden, synthetic property named `WarningLevelOverridden`. When it is true, we replace the `WarningLevel` controls with an equivalent read-only version and point the user to the `AnalysisLevel` property. A property interceptor computes the value of `WarningLevelOverridden` based on `EffectiveAnalysisLevel`.

This change also expands the set of admissible/supported `WarningLevel` values by adding those currently set by the .NET 6.0 SDK. This partially addresses the problem that we won't show the current value in the control if it isn't one of the admissible values, and may subsequently try to set the value to `null` when manipulating the value.

WarningLevelBehavior

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8004)

### CODE REVIEW

**Comment by tmeschter** [comment]
@drewnoakes I was planning on waiting until #7982 goes in and the rebasing on top of that.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/WarningLevelOverriddenValueProvider.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/WarningLevelOverriddenValueProvider.cs:38]: nit: use `decimal` over `float`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/WarningLevelOverriddenValueProvider.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/WarningLevelOverriddenValueProvider.cs:14]: There is a downside to coupling the implementation policy here to that in the SDK. In future, as the SDK changes, it can create issues when pinning the SDK.

Ideally the SDK would expose this information to the project system, as is done for target frameworks today. Older SDKs wouldn't provide that info, so we'd need some fallback.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/BuildPropertyPage.xaml
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/BuildPropertyPage.xaml:186]: Setting `Visible="False"` on the property directly is cleaner and runs slightly faster.

**Review by drewnoakes — APPROVED** [review]
> @drewnoakes I was planning on waiting until https://github.com/dotnet/project-system/pull/7982 goes in and the rebasing on top of that.

That PR is waiting for your approval. I think all your concerns were addressed there. You can merge it if you like.

Also feel free to close that PR and apply the changes you'd like to see here instead. That might be less work overall.

**Comment by melytc** [comment]
I might have got confused, but the idea here is that the only way a user targeting .NET5/6 can _actually_ decide the `WarningLevel` is by editing the `AnalysisLevel` property in Code Analysis? Because that ultimately affects that value with the `EffectiveAnalysisLevel`.

Also, would be great to have this documented in Microsoft Docs for when someone wonders why it changes to a read-only property. Thanks for the thoroughness!

**Comment by melytc** [comment]
I can open this as a separate issue, but could you check if the property is varied by configuration is reflected correctly? I saw this when testing the past experience:

warninglevelbug

**Review by melytc — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/BuildPropertyPage.xaml
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/PropertyPages/BuildPropertyPage.xaml:186]: Right. Forgot about that one!

**Comment by tmeschter** [comment]
> > @drewnoakes I was planning on waiting until #7982 goes in and the rebasing on top of that.
> 
> That PR is waiting for your approval. I think all your concerns were addressed there. You can merge it if you like.

Ack. Sorry, I thought I had. I'll approve and merge that.

**Comment by tmeschter** [comment]
> I might have gotten confused, but the idea here is that the only way a user targeting .NET5/6 can _actually_ decide the `WarningLevel` is by editing the `AnalysisLevel` property in Code Analysis? Because that ultimately affects that value with the `EffectiveAnalysisLevel`.

Yes, that's correct, with one exception. If the `AnalysisLevel` is set to `none` then `WarningLevel` will default to 4, _but_ any value set in the project file will take precedence.

> Also, would be great to have this documented in Microsoft Docs for when someone wonders why it changes to a read-only property. Thanks for the thoroughness!

Yeah, I'll file a documentation bug about this.

**Pull request merged**
